File: app/src/main/java/com/example/myapplication/ScientificFragment.java
package com.example.myapplication;

import android.app.AlertDialog;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.fragment.app.Fragment;
import androidx.lifecycle.ViewModelProvider;

import com.example.myapplication.databinding.FragmentScientificBinding;
import com.example.myapplication.repository.HistoryRepository;
import com.example.myapplication.viewmodel.ScientificViewModel;

/**
 * UI-only layer. All calculation logic is delegated to ScientificViewModel + CalculationRepository.
 */
public class ScientificFragment extends Fragment {

    private FragmentScientificBinding binding;
    private ScientificViewModel viewModel;
    private boolean degreeMode = false;
    private String lastResult = "";

    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        try {
            binding = FragmentScientificBinding.inflate(inflater, container, false);
            return binding.getRoot();
        } catch (Exception e) {
            TextView tv = new TextView(inflater.getContext());
            tv.setText("加载失败: " + e.getClass().getSimpleName());
            tv.setTextColor(0xFFFFFFFF);
            tv.setBackgroundColor(0xFF1C1C1E);
            tv.setGravity(17);
            return tv;
        }
    }

    public void onViewCreated(@NonNull View view, Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        if (binding == null) return;

        try {
            viewModel = new ViewModelProvider(this).get(ScientificViewModel.class);

            // Observe results
            viewModel.getResult().observe(getViewLifecycleOwner(), result -> {
                if (result != null && result.isSuccess()) {
                    String s = viewModel.formatResult(result);
                    binding.tvResult.setText("= " + s);
                    lastResult = s;
                }
            });

            viewModel.getHistory().observe(getViewLifecycleOwner(), history -> {
                // history is auto-saved; no UI update needed here
            });

            viewModel.getErrorMessage().observe(getViewLifecycleOwner(), msg -> {
                if (msg != null) binding.tvResult.setText("错误: " + msg);
            });

            viewModel.isLoading().observe(getViewLifecycleOwner(), loading -> {
                if (loading) binding.tvResult.setText("计算中...");
            });

            // Prevent system keyboard
            binding.etExpression.setShowSoftInputOnFocus(false);
            binding.etExpression.setOnTouchListener((v, event) -> { v.onTouchEvent(event); return true; });
            // Update LaTeX preview as user types
            binding.etExpression.addTextChangedListener(new android.text.TextWatcher() {
                public void beforeTextChanged(CharSequence s, int start, int count, int after) {}
                public void onTextChanged(CharSequence s, int start, int before, int count) {}
                public void afterTextChanged(android.text.Editable s) {
                    binding.latexPreview.setExpression(s.toString());
                }
            });

            // Inflate keyboard
            ViewGroup kbc = (ViewGroup) view.findViewById(R.id.kb_container);
            if (kbc != null) {
                View kbv = getLayoutInflater().inflate(R.layout.layout_math_keyboard, kbc, true);
                MathKeyboardHelper kb = new MathKeyboardHelper(kbv, text -> binding.etExpression.append(text));
                kb.getEqualsButton().setOnClickListener(v -> calculate());
                kb.getBackspaceButton().setOnClickListener(v -> {
                    String t = binding.etExpression.getText().toString();
                    if (!t.isEmpty()) binding.etExpression.setText(t.substring(0, t.length() - 1));
                });
                kb.getAcButton().setOnClickListener(v -> {
                    binding.etExpression.setText("");
                    binding.tvResult.setText("");
                });
            }

            // Toolbar
            updateDegDisplay();
            binding.btnDeg.setOnClickListener(v -> {
                degreeMode = !degreeMode;
                updateDegDisplay();
            });
            binding.btnAns.setOnClickListener(v -> {
                if (!lastResult.isEmpty()) binding.etExpression.append(lastResult);
            });
            binding.btnAns.setOnLongClickListener(v -> { showHistory(); return true; });
            binding.btnComplexI.setOnClickListener(v -> binding.etExpression.append("i"));
            binding.btnPi.setOnClickListener(v -> binding.etExpression.append("pi"));
            binding.btnCopy.setOnClickListener(v -> {
                String txt = binding.etExpression.getText().toString();
                if (!txt.isEmpty()) {
                    android.content.ClipboardManager cm = (android.content.ClipboardManager)
                        requireContext().getSystemService(android.content.Context.CLIPBOARD_SERVICE);
                    cm.setPrimaryClip(android.content.ClipData.newPlainText("expr", txt));
                    binding.tvResult.setText("已复制");
                }
            });
            binding.btnPaste.setOnClickListener(v -> {
                android.content.ClipboardManager cm = (android.content.ClipboardManager)
                    requireContext().getSystemService(android.content.Context.CLIPBOARD_SERVICE);
                if (cm.hasPrimaryClip()) {
                    CharSequence pasted = cm.getPrimaryClip().getItemAt(0).getText();
                    if (pasted != null) binding.etExpression.append(pasted);
                }
            });

        } catch (Exception e) {
            binding.tvResult.setText("初始化错误: " + e.getClass().getSimpleName());
        }
    }

    private void updateDegDisplay() {
        binding.btnDeg.setText(degreeMode ? "DEG" : "RAD");
        binding.tvModeIndicator.setText(degreeMode ? "DEG" : "RAD");
        binding.tvModeIndicator.setTextColor(degreeMode ? 0xFFFF9F0A : 0xFF888888);
    }

    private void calculate() {
        String expr = binding.etExpression.getText().toString();
        if (expr.isEmpty()) return;
        viewModel.calculate(expr, degreeMode);
    }

    private void showHistory() {
        java.util.List<HistoryRepository.Entry> items = viewModel.getHistory().getValue();
        if (items == null || items.isEmpty()) {
            binding.tvResult.setText("暂无历史记录");
            return;
        }
        CharSequence[] lines = new CharSequence[Math.min(items.size(), 50)];
        for (int i = 0; i < lines.length; i++) {
            HistoryRepository.Entry e = items.get(i);
            lines[i] = e.getExpression() + " = " + e.getResult();
        }
        new AlertDialog.Builder(getContext())
            .setTitle("历史记录 (长按清空)")
            .setItems(lines, (dialog, which) -> {
                binding.etExpression.setText(items.get(which).getExpression());
            })
            .setPositiveButton("清空", (d, w) -> viewModel.clearHistory())
            .show();
    }

    @Override
    public void onDestroyView() { super.onDestroyView(); binding = null; }
}

File: app/src/main/java/com/example/myapplication/ScientificFragment.kt
package com.example.myapplication

import android.app.AlertDialog
import android.content.ClipData
import android.content.ClipboardManager
import android.content.Context
import android.os.Bundle
import android.text.Editable
import android.text.TextWatcher
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.TextView
import androidx.core.content.ContextCompat
import androidx.fragment.app.Fragment
import androidx.lifecycle.ViewModelProvider
import com.example.myapplication.databinding.FragmentScientificBinding
import com.example.myapplication.viewmodel.ScientificViewModel

/**
 * Modernized UI Layer in Kotlin.
 */
class ScientificFragment : Fragment() {

    private var _binding: FragmentScientificBinding? = null
    private val binding get() = _binding!!
    
    private lateinit var viewModel: ScientificViewModel
    private var degreeMode = false
    private var lastResult = ""

    override fun onCreateView(
        inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?
    ): View {
        return try {
            _binding = FragmentScientificBinding.inflate(inflater, container, false)
            binding.root
        } catch (e: Exception) {
            TextView(context).apply {
                text = "加载失败: ${e.javaClass.simpleName}"
                setTextColor(0xFFFFFFFF.toInt())
                setBackgroundColor(0xFF1C1C1E.toInt())
                gravity = 17
            }
        }
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        viewModel = ViewModelProvider(this)[ScientificViewModel::class.java]

        setupObservers()
        setupInput()
        setupToolbar()
        setupKeyboard(view)
    }

    private fun setupObservers() {
        viewModel.result.observe(viewLifecycleOwner) { result ->
            if (result != null && result.isSuccess) {
                val s = viewModel.formatResult(result)
                binding.tvResult.text = "= $s"
                lastResult = s
            }
        }

        viewModel.errorMessage.observe(viewLifecycleOwner) { msg ->
            msg?.let { binding.tvResult.text = "错误: $it" }
        }

        viewModel.isLoading.observe(viewLifecycleOwner) { loading ->
            if (loading) binding.tvResult.text = "计算中..."
        }

        viewModel.degreeMode.observe(viewLifecycleOwner) { isDeg ->
            degreeMode = isDeg
            updateDegDisplay()
        }
    }

    private fun setupInput() {
        binding.etExpression.apply {
            showSoftInputOnFocus = false
            setOnTouchListener { v, event -> 
                v.onTouchEvent(event)
                true 
            }
            addTextChangedListener(object : TextWatcher {
                override fun beforeTextChanged(s: CharSequence?, start: Int, count: Int, after: Int) {}
                override fun onTextChanged(s: CharSequence?, start: Int, before: Int, count: Int) {}
                override fun afterTextChanged(s: Editable?) {
                    binding.latexPreview.setExpression(s.toString())
                }
            })
        }
    }

    private fun setupToolbar() {
        updateDegDisplay()
        
        binding.btnDeg.setOnClickListener {
            viewModel.toggleDegreeMode()
        }
        
        binding.btnAns.setOnClickListener {
            if (lastResult.isNotEmpty()) binding.etExpression.append(lastResult)
        }
        
        binding.btnAns.setOnLongClickListener {
            showHistory()
            true
        }

        binding.btnCopy.setOnClickListener {
            val txt = binding.etExpression.text.toString()
            if (txt.isNotEmpty()) {
                val cm = requireContext().getSystemService(Context.CLIPBOARD_SERVICE) as ClipboardManager
                cm.setPrimaryClip(ClipData.newPlainText("expr", txt))
                binding.tvResult.text = "已复制"
            }
        }

        binding.btnPaste.setOnClickListener {
            val cm = requireContext().getSystemService(Context.CLIPBOARD_SERVICE) as ClipboardManager
            cm.primaryClip?.getItemAt(0)?.text?.let {
                binding.etExpression.append(it)
            }
        }
        
        // Simple symbols
        binding.btnComplexI.setOnClickListener { binding.etExpression.append("i") }
        binding.btnPi.setOnClickListener { binding.etExpression.append("pi") }
    }

    private fun setupKeyboard(view: View) {
        binding.kbView.apply {
            onInput = { text ->
                binding.etExpression.append(text)
            }
            onBackspace = {
                val t = binding.etExpression.text.toString()
                if (t.isNotEmpty()) binding.etExpression.setText(t.dropLast(1))
            }
            onClear = {
                binding.etExpression.setText("")
                binding.tvResult.text = ""
            }
            onEquals = {
                calculate()
            }
        }
    }

    private fun updateDegDisplay() {
        binding.btnDeg.text = if (degreeMode) "DEG" else "RAD"
        binding.tvModeIndicator.text = if (degreeMode) "DEG" else "RAD"
        binding.tvModeIndicator.setTextColor(
            if (degreeMode) 0xFFFF9F0A.toInt() else 0xFF888888.toInt()
        )
    }

    private fun calculate() {
        val expr = binding.etExpression.text.toString()
        if (expr.isEmpty()) return
        viewModel.calculate(expr, degreeMode)
    }

    private fun showHistory() {
        val items = viewModel.history.value ?: return
        if (items.isEmpty()) {
            binding.tvResult.text = "暂无历史记录"
            return
        }

        val lines = items.take(50).map { "${it.expression} = ${it.result}" }.toTypedArray()
        
        AlertDialog.Builder(context)
            .setTitle("历史记录 (长按清空)")
            .setItems(lines) { _, which ->
                binding.etExpression.setText(items[which].expression)
            }
            .setPositiveButton("清空") { _, _ -> viewModel.clearHistory() }
            .show()
    }

    override fun onDestroyView() {
        super.onDestroyView()
        _binding = null
    }
}
